File: Algorithms/Easy/110_observation_students/solution.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int main() {
    int n, m;
    cin >> n >> m;
    vector<pair<int, int>> events;
    
    for (int i = 0; i < m; ++i) {
        int b1, b2;
        cin >> b1 >> b2;
        events.emplace_back(b1, 0);
        events.emplace_back(b2, 1);
    }
    
    events.emplace_back(n + 1, 0);
    events.emplace_back(n + 1, 1);
    
    sort(events.begin(), events.end());
    
    int prev = 0;
    int s = 0;
    int active_segments = 0;
    
    for (const auto& event : events) {
        int point = event.first;
        int event_type = event.second;
        
        if (active_segments == 0) {
            s += point - prev - 1;
        }
        if (event_type == 1) {
            prev = point;
            active_segments -= 1;
        } else {
            active_segments += 1;
        }
    }
    
    cout << s << endl;

    return 0;
}

File: Algorithms/Easy/110_observation_students/solution.py
def main():
    n, m = map(int, input().split())  # Read n and m from input
    events = []
    
    for _ in range(m):
        start, end = map(int, input().split())  # Read each pair of b1 and b2
        events.append((start, 0))  # Add start event
        events.append((end, 1))  # Add end event
    
    events.append((n + 1, 0))  # Add a dummy start event at n+1
    events.append((n + 1, 1))  # Add a dummy end event at n+1
    
    events.sort()  # Sort events by point value
    
    prev = 0  # Initialize previous point
    result = 0  # Initialize sum of uncovered segments
    balance = 0  # Initialize count of active segments
    
    for point, event_type in events:
        if balance == 0:  # If no active segments
            result += point - prev - 1  # Add the length of the uncovered segment
        if event_type == 1:  # If it's an end event
            prev = point  # Update previous point
            balance -= 1  # Decrease active segments count
        else:  # If it's a start event
            balance += 1  # Increase active segments count
    
    print(result)  # Print the result

if __name__ == "__main__":
    main()